### PR TITLE
[CSUB-1011] EVM Tracer to CC3

### PR DESCRIPTION
**Pull request by frankli-dev**
# Description of proposed changes

<describe what this PR is about and why we want it>

---

Practical tips for PR review & merge:

- [ ] All GitHub Actions report PASS
- [ ] Newly added code/functions have unit tests
  - [ ] Coverage tools report all newly added lines as covered
  - [ ] The positive scenario is exercised
  - [ ] Negative scenarios are exercised, e.g. assert on all possible errors
  - [ ] Assert on events triggered if applicable
  - [ ] Assert on changes made to storage if applicable
- [ ] Modified behavior/functions - try to make sure above test items are covered
- [ ] Integration tests are added if applicable/needed

### CODE REVIEW

File: runtime/Cargo.toml [hunk 1]
```diff
@@ -74,6 +74,12 @@ pallet-evm-precompile-sha3fips = { workspace = true }
 pallet-evm-precompile-simple = { workspace = true }
 pallet-hotfix-sufficients = { workspace = true }
 
+# Moonbeam tracing
+evm-tracing-events = { workspace = true, optional = true }
+moonbeam-evm-tracer = { workspace = true, optional = true }
+moonbeam-rpc-primitives-debug = { workspace = true }
+moonbeam-rpc-primitives-txpool = { workspace = true }
```
Comment (atodorov) [inline runtime/Cargo.toml:81]: IMO we should add all of these Cargo.toml files into dependabot.yml in case there's ever a dependency which doesn't come from the workspace.

**Review by atodorov — CHANGES_REQUESTED** [review]
Failing CI jobs need some more attention.

File: runtime/src/lib.rs [hunk 2]
```diff
@@ -1441,6 +1442,132 @@ impl_runtime_apis! {
...
+
+    impl moonbeam_rpc_primitives_debug::DebugRuntimeApi<Block> for Runtime {
+        fn trace_transaction(
+            extrinsics: Vec<<Block as BlockT>::Extrinsic>,
+            traced_transaction: &EthereumTransaction,
+        ) -> Result<
+            (),
+            sp_runtime::DispatchError,
+        > {
+            #[cfg(feature = "evm-tracing")]
```
Comment (atodorov) [inline runtime/src/lib.rs:1478]: Question: is this going to be conditionally compiled or should be part of the regular build/docker image ? 

Maybe one option will be to remove the conditionals if we want to have this always enabled.

File: node/src/eth.rs [hunk 3]
```diff
@@ -112,13 +136,17 @@ pub trait EthCompatRuntimeApiCollection:
     sp_api::ApiExt<Block>
     + fp_rpc::ConvertTransactionRuntimeApi<Block>
     + fp_rpc::EthereumRuntimeRPCApi<Block>
+    + moonbeam_rpc_primitives_debug::DebugRuntimeApi<Block>
+    + moonbeam_rpc_primitives_txpool::TxPoolRuntimeApi<Block>
```
Comment (atodorov) [inline node/src/eth.rs:140]: These additional RPC calls would need some sort of integration test at the very minimum. 

Skeleton for running tests against blockchain added in https://github.com/gluwa/creditcoin3/pull/171, example RPC tests from Creditcoin2 -> https://github.com/gluwa/creditcoin/tree/dev/integration-tests/src/test/rpc

**Review by atodorov — CHANGES_REQUESTED** [review]

File: node/cli-opt/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    pub ethapi_max_permits: u32,
+    pub ethapi_trace_max_count: u32,
+    pub ethapi_trace_cache_duration: u64,
+    pub eth_log_block_cache: usize,
+    pub eth_statuses_cache: usize,
+    pub fee_history_limit: u64,
+    pub max_past_logs: u32,
+    pub relay_chain_rpc_urls: Vec<url::Url>,
+    pub tracing_raw_max_memory_usage: usize,
+    pub frontier_backend_config: FrontierBackendConfig,
```
Comment (nathanwhit) [inline node/cli-opt/src/lib.rs:110]: Not needed

File: primitives/account/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,48 @@
+[package]
```
Comment (nathanwhit) [inline primitives/account/Cargo.toml:1]: Crate not needed

File: primitives/core/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,21 @@
+[package]
```
Comment (nathanwhit) [inline primitives/core/Cargo.toml:1]: Crate not needed

**Review by nathanwhit — COMMENTED** [review]
I haven't thoroughly reviewed all of the vendored code (it's a lot of code), but the changes here look how I expected them to. There are some unnecessary changes/vendored code, and we should probably rename some of the packages from `moonbeam-*` to `creditcoin3-*` but largely looks good.

File: runtime/src/lib.rs (hunk 2 above)
Comment (nathanwhit) [reply runtime/src/lib.rs:1453]: It is conditionally compiled, since according to moonbeam the tracing functionality has a _lot_ of overhead (in execution time and memory usage) - so not all nodes will run with it. Only some nodes (the ones used by blockscout, for instance) will use an image built with this flag.

File: node/src/eth.rs (hunk 3 above)
Comment (atodorov) [reply node/src/eth.rs]: @nathanwhit https://github.com/gluwa/creditcoin3/tree/dev/cli/src/test/blockchain-tests is where blockchain tests live in this repo. The "integration-tests" directory is integration tests for the cli.

**Review by zacharyfrederick — DISMISSED** [review]

File: cli/src/test/blockchain-tests/helpers.ts [hunk 7]
```diff
@@ -0,0 +1,26 @@
+import { ethers } from 'ethers';
+import { readFile } from 'fs/promises';
+import path from 'path';
+
+export const deployContract = async (
```
Comment (atodorov) [inline cli/src/test/blockchain-tests/helpers.ts:5]: CC @pLabarta FYI, you may be looking into this soon.

**Comment by frankli-dev** [comment]
> Requested some changes, mainly around portability & readability of the newly added test.
> 
> The test logic itself looks good, however it fails with
> 
> ```
>     could not coalesce error (error={ "code": -32601, "message": "Method not found" }, payload={ "id": 59, "jsonrpc": "2.0", "method": "debug_traceTransaction"
> ```
> 
> See: [https://github.com/gluwa/creditcoin3/actions/runs/7627194788 /job/20776099144?pr=169](https://github.com/gluwa/creditcoin3/actions/runs/7627194788/job/20776099144?pr=169)
> 
> There are unresolved CI failures which need to be addressed.

I see why it's failing.
We should build the node with the additional flag.

```
--features=evm-tracing
```

Also, we should run the node with the additional flag.
```
--ethapi=debug,trace,txpool
```

@atodorov Could you please help me to add these to CI building & running stuff?

File: runtime/Cargo.toml (hunk 1 above)
Comment (atodorov) [reply runtime/Cargo.toml]: @frankli-dev, @nathanwhit do you want to address this one here or separately ?

**Comment by nathanwhit** [comment]
Spoke with Dave and agreed we'll merge this today without the integration test, and open a follow-up PR with it. I'm going to just roll back the branch (force push) and re-open a PR with the test.
